### ISSUE TITLE
Upstream HTTP filters

### ISSUE THREAD

**Issue by snowp** [envoyproxy/envoy#10455]
Today there is no way for an extension writer to access the individual requests that are sent by the router (or any other HTTP client). We should extend the API to support adding upstream HTTP filters: same interface as the existing HTTP filters, but they apply to outbound requests. 

Originally I envisioned these to be on the router (similar to how the upstream logs are defined there), but perhaps it makes the most sense to put them on the cluster level so that they can apply to any HTTP call made through Envoy? Thinking about calls made through the AsyncClient

**Comment by wbpcode** [comment]
But clusters in envoy do not distinguish between protocols. In other words, a cluster may have redis/dubbo/thrift/http as its backend. If an http filter is added to the cluster, we need to make an additional distinction for each cluster.

**Comment by snowp** [comment]
We already have HTTP specific configuration on the cluster, so I don't think this is introducing any new protocol awareness to the cluster. That said, it might make sense to make it a bit more generic, i.e. something like

```
message ProtocolFilters {
  repeated google.protobuf.Any filters = 1;
}

map<string, ProtocolFilters> upstream_protocol_fillters = 1;
```

so that we don't have to add more HTTP specific fields to the cluster. At that point other L7 filters could use the same field to configure this. Alternatively we could use the metadata field, but this might be generic enough that having it be its own field makes sense.

**Comment by dio** [comment]
@snowp this is different from https://github.com/envoyproxy/envoy/blob/4403003b9c0888ce81606a2f45fcc1da126514e7/api/envoy/api/v2/cluster.proto#L776-L779? I can see there is one "interpretation" of it here: https://github.com/envoyproxy/envoy/issues/10525.

**Comment by snowp** [comment]
Those are L4 filters applied to the upstream connection, so not protocol aware.

**Comment by Augustyniak** [comment]
At Lyft, we would benefit from being able to access individual requests that are sent by the router too. I explained our use case in detail in [EnvoyMobile repository issue](https://github.com/lyft/envoy-mobile/issues/782) but Envoy repository may be a better place for an issue like this. 

TL;DR;
`EnvoyMobile` is a version of `Envoy` intended to be run on mobile devices. Client-side networking frameworks such as `URLSession` or `Alamofire` allow for access to individual requests that they make. Our transition to `EnvoyMobile` requires some workarounds because with `Envoy` (and as a result with `EnvoyMobile` too) it's not possible to access all requests that the router makes. See https://github.com/lyft/envoy-mobile/issues/782 for more details.

**Comment by mattklein123** [comment]
@Augustyniak for Envoy Mobile I don't think you need upstream HTTP filters since you have access to every downstream request as a normal filter? Can you describe the use case where you need an upstream HTTP filter? cc @junr03

**Comment by junr03** [comment]
The particular use case I discussed with @Augustyniak on friday was being able to set fault injection headers independently for each retry in a series of retries, rather than all retries having the same headers set when the downstream request passes through the current http fault filter. More of @Augustyniak's thoughts are in https://github.com/lyft/envoy-mobile/issues/782

We discussed some shorter term solutions they can pursue (like controlling faults in the edge proxy). However, longer term (for this particular use case, as well as potentially other client resilience projects) the idea of filters that can act on _every_ upstream request is pretty powerful.

@Augustyniak let us know if I missed something.

**Comment by mattklein123** [comment]
> However, longer term (for this particular use case, as well as potentially other client resilience projects) the idea of filters that can act on every upstream request is pretty powerful.

Yes, thanks, makes sense.

**Comment by Augustyniak** [comment]
@mattklein123 @junr03's covers is well. I just posted another example in https://github.com/lyft/envoy-mobile/issues/782 for when being able to update upstream requests Envoy makes could be helpful.

TL;DR;
Our mobile applications attach `x-timestamp-ms` HTTP header to every upstream request they make. This HTTP header is supposed to contain the NTP timestamp as measured on the client at the time of the upstream request start. With `EnvoyMobile` (and `Envoy` too) we can only set the value of `x-timestamp-ms` once for when a downstream request is started and have no way of updating it every time a new upstream request is started.

**Comment by snowp** [comment]
Now that `FilterManager` has been moved out of HCM, I've started looking at what this will require. Some notes:

- FilterManager currently takes ownership over the headers/trailers - this won't work with the upstream case because the headers will be owned by the HCM and we need to keep it that way (due to access logging etc.). Easiest solution here is to
have the FM keep non-owning pointers to the headers/trailers.
- The naive solution would expose the same encoder/decoder callbacks for upstream filters as for the existing HTTP filters, but some functions might not make sense: for example all the callback functions relating to the route resolution. The easiest solution here would be to just blow up if these functions are called, but we might want to consider templating FM to allow different callback classes to be used. 
- As a first step we can simply have the router use the FM with a single filter that handles the upstream stream lifecycle, completely reset per attempt. Then down the line we can move the retry semantics into a filter that utilizes `recreateStream` to issue retry attempts

fyi @alyssawilk in case you have any thoughts here

**Comment by alyssawilk** [comment]
scattered thoughts :-)

If we change the ownership model we need to be super careful about owner and lifetime. We've already seen one crashing bug for a corner case like original response headers being overwritten by a filter sendLocalReply or some such

Prefer template over hard fail (esp in prod) but launch and iterate is 100% fine, esp in alpha.

I assume for hedging we're going to want a upstream FM per hedge attempt.  I would have thought we would want similar for retries, where the upstream filter was only used for one attempt so it only had to deal with decoding headers once for example.  Are you just saying that the upstream filter can force a retry via recreate stream, or that it would also receive the retry?

**Comment by snowp** [comment]
Changing the ownership model is definitely tricky, I'm currently playing with it to see if I can make sense of it :) 

Re retries: basically I was trying to frame the retry behavior in terms of existing functionality that the FM already has. I think each attempt should have its own FM, with state shared between attempts using the FilterState object. This would facilitate stuff like adjusting headers on subsequent retry attempts and other per attempt adjustments. 

I think retries would be modeled by `recreateStream` while hedges would use `cloneStream`/`cloneAndRestartStream` or something to that effect, the first one ending the current stream while the other keeps the current stream alive. 

The nice thing about this is that it would make it a lot easier for users to add their own hedging/retry implementations, as they would have full control over retry/hedge attempts in their filters.

**Comment by lambdai** [comment]
The ownership is too hard for me to understand.

I want to add the use case that upstream filter manager should be able to reflect the latest host after retry.

This feature is wanted
https://github.com/envoyproxy/envoy/issues/12236
https://github.com/envoyproxy/envoy/issues/13897

And currently cluster upstream extension could be used to rewrite header after HCM router filter rewrite the header.
See https://github.com/envoyproxy/envoy/pull/13915

**Comment by snowp** [comment]
To support this we'd either want all filters to run after host selection (this precludes some features that would want pre-host selection filters) or have filters run during both. For the latter we could either have two lists of filters or have a host selection filter that can be inserted in the middle to handle selection. I don't think any work has been done so far to exclude either option, this will likely come into play once we start talking about the API. 

Open to input on the best approach here, but my focus is elsewhere so I'm not expecting to make much progress here in the short term.

**Comment by markdroth** [comment]
We've run into a use-case in gRPC that requires something like this.  We need to configure fault-injection on a per-locality basis.  This requires running the fault injection filter after load balancing is complete, so that we know what locality the request is going to be routed to and can use that as input to decide whether to inject a fault.

What is the current state of this work?  Note that for our use-case, we don't need this feature to actually be implemented in Envoy, but we do need to get the relevant xDS API fields defined and come to agreement on the exact semantics, so that we can proceed with implementation in gRPC.

**Comment by markdroth** [comment]
Just thinking out loud here, based on a discussion I had on the gRPC side with @ejona86 and @dfawley:

Looking at the list of use-cases described above, it seems like most of them require only intercepting the start of an upstream request (i.e., headers), not looking at the payload of the request.  Do we know of any use-cases that would require looking at the request payloads?

If all of the expected use-cases only actually need to look at headers, do we actually want to provide a full HTTP filter API for this, or would it make sense to instead provide a more constrained API that supports only looking at headers?  That might be easier to implement in both Envoy and gRPC, and exposing less of an API surface may provide more flexibility for implementations.  On the flip side, though, exposing a different API on the upstream side would mean that it's not possible to move existing filters (like the fault injection filter) from the HCM to the upstream side without substantial changes.

Thoughts?

**Comment by wbpcode** [comment]
🤔  If we don't consider the route entry refresh, then when the requests are coming, we can simple create a filter chain which contains downstream filters and upstream filters for the requests.

For example, `downstream lua` -> `downstream fault` -> `upstream lua` -> `upstream fault` -> `router`.

It would be very easy to implement.

**Comment by alyssawilk** [comment]
I think we're going to want payload filters in the long run, in case say some cluster doesn't support compressed data and we want to decompress to a given cluster.  And I think it's nice to be able to write a filter once and reuse it in multiple places (route refresh aside)

**Comment by wbpcode** [comment]
@alyssawilk is there any related proposal for this feature? or is anyone working on this?

**Comment by alyssawilk** [comment]
Snow did about half the work a few years back.  I was toying around with picking it up this quarter - was looking last week to see how horrendous it'd be.  I'll probably sign on for some of the clean up refactoring at least so we're in a better position to land it.

**Comment by markdroth** [comment]
@alyssawilk 
> I think we're going to want payload filters in the long run, in case say some cluster doesn't support compressed data and we want to decompress to a given cluster. And I think it's nice to be able to write a filter once and reuse it in multiple places (route refresh aside)

Okay, that's kind of what I was figuring the answer would be.  In that case, I think it does make sense for the xDS API to actually be a full HTTP filter, using the same API that we use in the HCM.

@wbpcode 
> If we don't consider the route entry refresh, then when the requests are coming, we can simple create a filter chain which contains downstream filters and upstream filters for the requests.
> 
> For example, `downstream lua` -> `downstream fault` -> `upstream lua` -> `upstream fault` -> `router`.
> 
> It would be very easy to implement and all the current filters can be used as upstream HTTP filters directly.

I don't think it will work for the upstream filters to be invoked before the router filter, because some of these use-cases require the upstream filters to run after load balancing is done, so that the load balancing decision can be used by the upstream filter to determine behavior (e.g., enabling fault injection based on which locality the request has been routed to).

**Comment by wbpcode** [comment]
@markdroth Yeah, it has lots of shortages except it is easy to implement.

It would be very complex to create additional filter chain for upstream HTTP filters. I didn't come up a good way to manage this new filter chain.

Or we can try to split router to two different parts. One for load balancing, this one will be called before upstream filters. One for sending upstream request, at the end of whole filter chain. But it's weird 🤣.

**Comment by markdroth** [comment]
@wbpcode I think being easy to implement isn't actually an advantage if it doesn't solve many of the use cases that we're considering here. :)  If you look at the use-cases described above, there are also several that require invoking the upstream filters separately for each call attempt done for retries or hedging, which I think also requires running them after the router filter.

I'll let folks who are more familiar with Envoy's architecture address the question of the right way to implement this.

**Comment by wbpcode** [comment]
@markdroth It would be perfect if we can get a independent and fully functional upstream HTTP filter chain.

But the state of a complete HTTP filter chain is just too complicated orz.

**Comment by kyessenov** [comment]
@wbpcode We have some use-cases where we could use upstream HTTP filters for CONNECT streams initiated by `tcp_proxy`. I think that prevents reliance on `router` for upstream HTTP filters.

**Comment by wbpcode** [comment]
> @wbpcode We have some use-cases where we could use upstream HTTP filters for CONNECT streams initiated by `tcp_proxy`. I think that prevents reliance on `router` for upstream HTTP filters.

This sounds reasonable scene. But as far as I know, We may implement a independent upstream filter chain in the router. The scene of `tcp_proxy` is not taken into  account. 

@alyssawilk is working on this. And we can wait the concrete proposal to be completed and make some more discussions.

**Comment by alyssawilk** [comment]
so @paul-r-gall brings up a design point I'd like folks thoughts on (cc @snowp @mattklein123 )

he took an existing filter we have internally and is running them as upstream mirrored filters.  His expectation was that the mirrored request would run the decoder paths but not encoder paths. I hadn't exempted the encode functions from anything since we've traditionally run the "full upstream code" for mirroring

It'd be pretty easy to only register decode callbacks for mirrored requests for upstream filters.  I just don't know if that should be The Way of doing things, or a configurable option.  Thoughts?

**Comment by mattklein123** [comment]
> It'd be pretty easy to only register decode callbacks for mirrored requests for upstream filters. I just don't know if that should be The Way of doing things, or a configurable option. Thoughts?

Personally I would always run decode/encode as filters may have assumptions about what is done in these callbacks. We would just eat the final encode for the mirror. I suppose we could make this configurable also but seems better to avoid that unless someone really wants it?

**Comment by mattklein123** [comment]
> For mirrored requests, the encode function for downstream http filters is not run. This is why I was surprised that encode was run for upstream http filters on mirrored requests.

I see. I think that is an artifact of mirror requests being not a a real filter chain (something that has been commented on and would be nice to fix but would be a huge undertaking). IMO running the encode callbacks is more correct, but I don't have a strong opinion on what we do here.

**Comment by alyssawilk** [comment]
I'm sure there will be more added here but I think this is done enough for trial use.

**Comment by michaelfinch** [comment]
Are upstream filters run after host selection? I'm wondering if I'll be able to access endpoint metadata from within an upstream filter.

**Comment by alyssawilk** [comment]
I don't think you have access to the host info guaranteed on filter creation as connection and so host selection is done in parallel, by default, but if you need the info you can have your filter pause,  wait for UpstreamCallbacks::onUpstreamConnectionEstablished (which you have to explicitly register for) and then continueDecoding

**Comment by michaelfinch** [comment]
Great, I confirmed that it works as you described. Thank you for adding this feature.
